### ISSUE TITLE
Kusama "RemainingScaleBytesNotEmptyException" when calling the `get_block` method

### ISSUE THREAD

**Issue by Oladiy** [polkascan/py-scale-codec#79]
Hi. When I call the `get_block` method of the py-substrate-interface library, an exception `RemainingScaleBytesNotEmptyException' occurs for specific block. The exception started to occur after a recent runtime update.

Network: `kusama`
Block number: `14133297`
Endpoint: `wss://kusama-rpc.polkadot.io`

substrate-interface: `v1.2.9`
scalecodec: `v1.0.41`

Logs:
```
   DEBUG 2022-08-31 17:33:57 <base> <daemon> RPC request #3: "chain_getBlockHash"
   DEBUG 2022-08-31 17:33:57 <base> <daemon> RPC request #4: "chain_getHeader"
   DEBUG 2022-08-31 17:33:57 <base> <daemon> RPC request #5: "chain_getRuntimeVersion"
   DEBUG 2022-08-31 17:33:57 <base> <daemon> RPC request #6: "state_getMetadata"
   DEBUG 2022-08-31 17:33:58 <base> <daemon> Retrieved metadata for 9260 from Substrate node
   DEBUG 2022-08-31 17:33:58 <base> <daemon> Add PortableRegistry from metadata to type registry
   DEBUG 2022-08-31 17:33:58 <base> <daemon> RPC request #7: "chain_getBlock"
   ERROR 2022-08-31 17:33:58 <start> <daemon> Error <class 'scalecodec.exceptions.RemainingScaleBytesNotEmptyException'>: Decoding <U64> - No more bytes available (needed: 194 / total: 188)
```

**Comment by Nadro-J** [comment]
Have you tried updating both substrate-interface and scalecodec?
`pip3 install substrate-interface scalecodec --upgrade`

**Comment by Oladiy** [comment]
> pip3 install substrate-interface scalecodec --upgrade

Yes, it has already been updated

**Comment by Nadro-J** [comment]
What does your code look like? The following below works for me on Ubuntu 21.04

|       | version | 
| ------------- |:-------------:| 
| python      | 3.9.5 | 
| scalecodec      | 1.0.41     | 
| substrate-interface | 1.2.9      | 


```python
from substrateinterface import SubstrateInterface

try:
    substrate = SubstrateInterface(
        url="wss://kusama-rpc.polkadot.io/"
    )
except ConnectionRefusedError:
    print("⚠️ Unable to connect!")
    exit()

blockInfo = substrate.get_block(block_number=14133297)
print(blockInfo)
```

**Comment by Oladiy** [comment]
> What does your code look like? The following below works for me on Ubuntu 21.04
> 
> version
> python	3.9.5
> scalecodec	1.0.41
> substrate-interface	1.2.9
> ```python
> from substrateinterface import SubstrateInterface
> 
> try:
>     substrate = SubstrateInterface(
>         url="wss://kusama-rpc.polkadot.io/"
>     )
> except ConnectionRefusedError:
>     print("⚠️ Unable to connect!")
>     exit()
> 
> blockInfo = substrate.get_block(block_number=14133297)
> print(blockInfo)
> ```

I used the same code and got the same exception. This code is executed inside the docker container.
|   | version |
| ------------- | ------------- |
| python | 3.9.13 |
| scalecodec  | 1.0.41  |
| substrate-interface | 1.2.9|
| Ubuntu | 20.04.01 |

When I try to execute
```
from substrateinterface import SubstrateInterface
substrate = SubstrateInterface(url="wss://kusama-rpc.polkadot.io")
substrate.update_type_registry_presets()
substrate.reload_type_registry()
substrate.get_block(block_number=14133297)
```
, an exception occurs:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/.local/lib/python3.9/site-packages/substrateinterface/base.py", line 2635, in get_block
    return self.__get_block_handler(
  File "/root/.local/lib/python3.9/site-packages/substrateinterface/base.py", line 2594, in __get_block_handler
    return decode_block(response['result']['block'], block_data_hash=block_hash)
  File "/root/.local/lib/python3.9/site-packages/substrateinterface/base.py", line 2504, in decode_block
    extrinsic_decoder.decode()
  File "/root/.local/lib/python3.9/site-packages/scalecodec/base.py", line 677, in decode
    self.value_serialized = self.process()
  File "/root/.local/lib/python3.9/site-packages/scalecodec/types.py", line 2567, in process
    self.value_object.update(self.process_type('Inherent', metadata=self.metadata).value_object)
  File "/root/.local/lib/python3.9/site-packages/scalecodec/base.py", line 760, in process_type
    obj.decode(check_remaining=False)
  File "/root/.local/lib/python3.9/site-packages/scalecodec/base.py", line 677, in decode
    self.value_serialized = self.process()
  File "/root/.local/lib/python3.9/site-packages/scalecodec/types.py", line 506, in process
    field_obj = self.process_type(data_type, metadata=self.metadata)
  File "/root/.local/lib/python3.9/site-packages/scalecodec/base.py", line 760, in process_type
    obj.decode(check_remaining=False)
  File "/root/.local/lib/python3.9/site-packages/scalecodec/base.py", line 677, in decode
    self.value_serialized = self.process()
  File "/root/.local/lib/python3.9/site-packages/scalecodec/types.py", line 1323, in process
    call_obj = self.process_type(call_type_string, metadata=self.metadata)
  File "/root/.local/lib/python3.9/site-packages/scalecodec/base.py", line 759, in process_type
    obj = self.runtime_config.create_scale_object(type_string, self.data, **kwargs)
  File "/root/.local/lib/python3.9/site-packages/scalecodec/base.py", line 168, in create_scale_object
    raise NotImplementedError('Decoder class for "{}" not found'.format(type_string))
NotImplementedError: Decoder class for "scale_info::202" not found
```

**Comment by arjanz** [comment]
Ok after some debugging I think I found the culprit. 

After I add 

```python
import logging
logging.basicConfig(level=logging.DEBUG)
```
to enable debugging, I saw the line `substrate.update_type_registry_presets()` triggered a few requests in order to update the local type registry presets. 

Although since introduction of MetadataV14 - which embedded a type registry - this is not necessary anymore, if you want to retrieve historic blocks or state pre-MetadataV14, this is still required. 

Then the retrieved URL rang a bell:

```
DEBUG:urllib3.connectionpool:https://raw.githubusercontent.com:443 "GET /polkascan/py-scale-codec/v1.0/scalecodec/type_registry/default.json HTTP/1.1" 200 23221
```

I haven't merged the changes of  `master ` into `v1.0` for a while, slipped my mind for the same reason this is not really necessary anymore. Unless you can the `substrate.update_type_registry_presets()` function. 

So I merged `master` into `v1.0` and locally the problem disappeared, I'm wondering if it is also working for you now @Oladiy ?

So a bit of a mea culpa here, I'll put this on the release checklist to make sure this doesn't happen again.

**Comment by Oladiy** [comment]
> Ok after some debugging I think I found the culprit.
> 
> After I add
> 
> ```python
> import logging
> logging.basicConfig(level=logging.DEBUG)
> ```
> 
> to enable debugging, I saw the line `substrate.update_type_registry_presets()` triggered a few requests in order to update the local type registry presets.
> 
> Although since introduction of MetadataV14 - which embedded a type registry - this is not necessary anymore, if you want to retrieve historic blocks or state pre-MetadataV14, this is still required.
> 
> Then the retrieved URL rang a bell:
> 
> ```
> DEBUG:urllib3.connectionpool:https://raw.githubusercontent.com:443 "GET /polkascan/py-scale-codec/v1.0/scalecodec/type_registry/default.json HTTP/1.1" 200 23221
> ```
> 
> I haven't merged the changes of `master ` into `v1.0` for a while, slipped my mind for the same reason this is not really necessary anymore. Unless you can the `substrate.update_type_registry_presets()` function.
> 
> So I merged `master` into `v1.0` and locally the problem disappeared, I'm wondering if it is also working for you now @Oladiy ?
> 
> So a bit of a mea culpa here, I'll put this on the release checklist to make sure this doesn't happen again.

It works now, thanks!
Also can you explain me the next thing? I created the SubstrateInterface object, tried to get the 14133297 block and got the same exception (`scalecodec.exceptions.RemainingScaleBytesNotEmptyException`). Then I called `update_type_registry_presets()` and got the `NotImplementedError: Decoder class for "scale_info::202" not found` exception. But after I recreated the SubstrateInterface object, the problem disappeared at all. 
Unfortunately, I am not able to share the logs, but the code was:
```
from substrateinterface import SubstrateInterface
substrate = SubstrateInterface(url="wss://kusama-rpc.polkadot.io/")
blockInfo = substrate.get_block(block_number=14133297)
# here is the `scalecodec.exceptions.RemainingScaleBytesNotEmptyException` exception
substrate.update_type_registry_presets()
blockInfo = substrate.get_block(block_number=14133297)
# here is the `NotImplementedError: Decoder class for "scale_info::202" not found` exception
substrate = SubstrateInterface(url="wss://kusama-rpc.polkadot.io/")
blockInfo = substrate.get_block(block_number=14133297)  # now it works
```
